src/structs/branch.rs

```diff
@@ -74,32 +74,78 @@ impl Branch {
     }
 
     #[inline(always)]
-    pub fn record_and_update_true_obs(&mut self) {
-        if (self.counts & 0xff) != 0xff {
-            // non-overflow case is easy
-            self.counts += 1;
-        } else {
-            // normalize, except special case where it is all trues
-            if self.counts != 0x01ff {
-                self.counts = (((self.counts as u32 + 0x100) >> 1) & 0xff00) as u16 | 129;
-            }
+    pub fn record_and_update_bit(&mut self, bit: bool) {
+        // rotation is used to update either the true or false counter
+        // this allows the same code to be used without branching,
+        // which makes the CPU about 20% happier.
+        //
+        // Since the bits are randomly 1/0, the CPU branch predictor does
+        // a terrible job and ends up wasting a lot of time. Normally
+        // branches are a better idea if the branch very predictable vs
+        // this case where it is better to always pay the price of the
+        // extra rotation to avoid the branch.
+        let orig = self.counts.rotate_left(bit as u32 * 8);
+        let (mut sum, o) = orig.overflowing_add(0x100);
+        if o {
+            // normalize, except in special case where we have 0xff or more same bits in a row
+            // in which case we want to bias the probability to get better compression
+            //
+            // CPU branch prediction soon realizes that this section is not often executed
+            // and will optimize for the common case where the counts are not 0xff.
+            let mask = if orig == 0xff01 { 0xff00 } else { 0x8100 };
+            sum = ((1 + (sum & 0xff)) >> 1) | mask;
```

Comment:
```suggestion
            sum = ((1 + sum) >> 1) | mask;
```

You have 0 in the upper byte anyway.

Reply:
ok done. thanks!